src/main/java/com/iota/iri/utils/dag/DAGUtils.java

```diff
@@ -0,0 +1,366 @@
+     *
+     * To make the use of this method as easy as possible we provide a broad range of possible parameter signatures, so
+     * we do not have to manually convert between milestones, transactions and hashes (see other methods with the same
+     * name).
+     *
+     * It uses an non-recursive iterative algorithm that is able to handle huge chunks of the tangle without running out
+     * of memory. It creates a queue of transactions that are being examined and processes them one by one. As new
+     * approvees are found, they will be added to the queue and processed accordingly.
+     *
+     * Every found transaction is passed into the provided condition lambda, to determine if it still belongs to the
+     * desired set of transactions and only then will be passed on to the currentTransactionConsumer lambda.
+     *
+     * @param startingTransaction the starting point of the traversal (does not get passed to the consumer)
+     * @param condition a lambda function that allows to control how long the traversal should continue
+     * @param currentTransactionConsumer a lambda function that allows us to "process" the found transactions
+     * @param processedTransactions a set of hashes that shall be considered as "processed" already and that will
+     *                              consequently be ignored in the traversal
+     * @throws TraversalException if anything goes wrong while traversing the graph and processing the transactions
+     */
+    public void traverseApprovees(TransactionViewModel startingTransaction,
+                                  TraversalCondition condition,
+                                  TraversalConsumer currentTransactionConsumer,
+                                  Set<Hash> processedTransactions) throws TraversalException {
+        final Queue<TransactionViewModel> transactionsToExamine = new LinkedList<>();
+        try {
+            transactionsToExamine.add(TransactionViewModel.fromHash(tangle, startingTransaction.getBranchTransactionHash()));
+            transactionsToExamine.add(TransactionViewModel.fromHash(tangle, startingTransaction.getTrunkTransactionHash()));
+
+            TransactionViewModel currentTransaction;
+            while((currentTransaction = transactionsToExamine.poll()) != null) {
+                if(
+                processedTransactions.add(currentTransaction.getHash()) &&
+                currentTransaction.getType() != TransactionViewModel.PREFILLED_SLOT &&
+                condition.check(currentTransaction)
+                ) {
+                    currentTransactionConsumer.consume(currentTransaction);
+
+                    transactionsToExamine.add(TransactionViewModel.fromHash(tangle, currentTransaction.getBranchTransactionHash()));
+                    transactionsToExamine.add(TransactionViewModel.fromHash(tangle, currentTransaction.getTrunkTransactionHash()));
```

Comment:
This implementation is BFS, any chance we might need a DFS one since we are at it?

Reply:
We can add more functions when we need them - for now BFS is the "fairest" when it comes to traversing subtangles because it doesnt prioritize a single path (and the most used one in the code).